File: python/coin-change.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        amounts = [amount + 1] * (amount + 1)
        amounts[0] = 0
        for j in range(1, amount + 1):
            for coin in coins:
                if j - coin >= 0:
                    amounts[j] = min(amounts[j], 1 + amounts[j-coin])
        return amounts[-1] if (amounts[-1] != amount + 1) else -1

File: rust/coin-change.rs
use std::cmp::{min};

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let max_value = amount + 1;
        let mut amounts = vec![max_value; max_value as usize];
        amounts[0] = 0;
        for i in 1..=amount {
            for coin in &coins {
                if i - coin >= 0 {
                    amounts[i as usize] = min(amounts[i as usize], 1 + amounts[(i - coin) as usize]);
                }
            }
        }
        if amounts[amount as usize] == max_value {
            -1
        } else {
            amounts[amount as usize]
        }
    }
}